portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,7 +34,62 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    // Get all messages stored on Datastore
+    ArrayList<String> messages = new ArrayList<String>();
+    for (Entity entity : results.asIterable()) {
+
+      String comment = (String) entity.getProperty("comment");
+      messages.add(comment);
+    }
+
     response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    String json = convertToJson(messages);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Parse input from the form
+    String comment = getParameter(request, "text-input", "");
+    boolean upperCase = Boolean.parseBoolean(getParameter(request, "upper-case", "false"));
+    boolean lowerCase = Boolean.parseBoolean(getParameter(request, "lower-case", "false"));
+    long timestamp = System.currentTimeMillis();
+
+    if (upperCase) {
+      comment = comment.toUpperCase();
+    }
+    if (lowerCase) {
+      comment = comment.toLowerCase();
+    }
+
+    Entity taskEntity = new Entity("Comment");
+    taskEntity.setProperty("comment", comment);
+    taskEntity.setProperty("timestamp", timestamp);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(taskEntity);
+
+    response.sendRedirect("/index.html");
+  }
+
+  private String convertToJson(ArrayList<String> messages) {
+    Gson gson = new Gson();
```

Comment:
nit: Same comment as on the Query object. This can be `private static final` since the Gson object does not seem to have any state that would need to be unique to each request. The functionality is the same it just cuts down on the number of memory allocations that have to happen when handling each request.

Reply:
Thank you for the information!